### ISSUE TITLE
How many times do I need to call model.zero_grad()?

### ISSUE THREAD

**Issue by gchhablani** [pytorch/captum#590]
Hi, 
I was trying to run IntegratedGradients for BERT (not LayerIntegratedGradients) based on this notebook: https://captum.ai/tutorials/Bert_SQUAD_Interpret

In the beginning, the author called `model.zero_grad()` and `model.eval()` once, while he continues to repeat LayerIntegratedGradients for both start and end, not doing `model.eval()` and `zero_grad()` in between.

My question is, are gradients being calculated in the network by LIG or IG? If yes, do I need to call `zero_grad()` and `eval()` everytime I run it.

Based on my understanding of IG, it is only based on input and the output, so is gradient being calculated at all?

**Comment by NarineK** [comment]
Hi @gchhablani, thank you for the question. Setting eval once for the model is sufficient. In terms of `zero_grad()` this is not mandatory since internally we take care of it. In the algorithms there zero-ing the grads is important we `apply_gradient_requirements()` function: https://github.com/pytorch/captum/blob/master/captum/_utils/gradient.py#L26
In case of IG and layer IG we don't really need to zero the grads because we recreate the inputs and for the new inputs grads are always zero.

**Comment by gchhablani** [comment]
Hi @NarineK,

Thanks for your prompt answer. I'll go through the code you shared in the above link.

I was wondering if that would changed if I were to do it for multiple samples, say, using a DataLoader? 
For example:
```
for i,batch in enumerate(data_loader):
    ig = IntegratedGradients(model)
   *inputs, labels = batch
   attributions,delta = ig.attribute(*inputs,target=labels, return_convergence_delta=True)
```
Thanks,
Gunjan

**Comment by bilalsal** [comment]
Hi Gunjan,

you can safely use both alternatives. The 2nd alternatives saves is how Captum is intended to be used: Define the attribution method ones, apply it as many times as needed.

As @NarineK mentioned, you do not need to worry about setting the gradient to zero even when using the 2nd alternative. Both `IntegratedGradients` or `LayerIntegratedGradients` operate by feeding _multiple versions_ of an input, interpolating from baseline to the one you provide. From the `model`'s point of view Each of these versions is an independent input, so its gradients will be recomputed. 

Hope this helps

**Comment by gchhablani** [comment]
Thanks a lot @bilalsal and @NarineK. Now, I understand it much better.

Closing this issue as my query has been aptly answered. :)
